src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -503,9 +500,8 @@ void Replayer<I>::handle_load_remote_group_snapshots(int r) {
 
 template <typename I>
 void Replayer<I>::validate_image_snaps_sync_complete(
-    const std::string &group_snap_id) {
-  std::unique_lock locker{m_lock};
-  if (is_replay_interrupted(&locker)) {
+    const std::string &group_snap_id, std::unique_lock<ceph::mutex>* locker) {
+  if (is_replay_interrupted(locker)) {
```

Comment:
I don't think this is required but the unlock and lock inside the function means that the is_replay_interrupted needs to be checked everytime the locker is locked as the shut_down could have set the state to COMPLETE if it gets the lock in that interval:

```
void Replayer<I>::shut_down(Context* on_finish) {
  dout(10) << dendl;
  
  {
    std::unique_lock locker{m_lock};
    m_stop_requested = false;
    ceph_assert(m_on_shutdown == nullptr);
    std::swap(m_on_shutdown, on_finish);
  
    auto state = STATE_COMPLETE;
    std::swap(m_state, state);
  }

```

Reply:
I have removed the check here and added checks right after taking the lock in the two other places in the routine.